### PR TITLE
Testing/rw test coverage

### LINKED ISSUE #58: App Testing
### Feature Categories:

- React
- Testing

**Describe the feature/Bug your gonna work on**
- Create tests for App component for integration & acceptance-level user stories

### PR DESCRIPTION
**Pull request by rwilliams659**
# Som Timer Pull Request Form

## Description
- App integration & acceptance-level tests were added for all possible user stories
- There are some limitations in testing due to lack of control on components imported from other libraries (React Player & React Compound Timer)
- Notification functions are broken out into their own file 
- Countdown Timer unit test added for calling the correct external functions when the timer runs out

**Note: With functional components that contain other sub-functions that get called onClick/on another action, these cannot be unit tested if they live within the function, only if they exist externally. All externally called functions are tested (i.e. the audio notification & display notification functions), but others are not.**

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Testing

## Has This Been Tested?
- [ ] No
- [X] Yes
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration:
Run `npm run test -- --coverage --watchAll=true` to see test coverage report 

# Checklist:
- [X] My code has no unused/commented out code
- [X] My code has no console logs
- [X] I have performed a self-review of my own code
- [X] I have commented on my code, particularly in hard-to-understand areas

resolve #58

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
     "clean-cra-sw": "rm -f build/precache-manifest.*.js && rm -f build/service-worker.js",
     "build": "react-scripts build && npm run build-sw && npm run clean-cra-sw",
     "test": "react-scripts test --env=jest-environment-jsdom-sixteen",
+    "test:coverage": "CI=true react-scripts test --env=jsdom",
```
Comment (sienna-kopf) [inline package.json:24]: yay! Test coverage tracking is the best!

File: src/Components/ContentSelection/ContentSelection.test.js [hunk 2]
```diff
@@ -20,25 +20,11 @@ describe('ContentSelection', () => {
...
-          </VideoProvider>
-        </ViewProvider>
-      </SettingsProvider>
-    )
-
     const yogaMovementButton = getByRole('button', { name: /yoga\/movement/i })
     const somaticExerciseButton = getByRole('button', { name: /somatic exercise/i })
     const breathworkMeditationButton = getByRole('button', { name: /breathwork\/meditation/i })
 
+    expect(contentPrompt).toBeInTheDocument()
```
Comment (sienna-kopf) [inline src/Components/ContentSelection/ContentSelection.test.js:27]: Nice addition here.

File: src/Components/CountdownTimer/CountdownTimer.test.js [hunk 3]
```diff
@@ -1,11 +1,26 @@
...
+import { displayNotification } from '../../helpers/notificationHelpers'
+import { fireEvent, render, waitFor } from '@testing-library/react'
+import '@testing-library/jest-dom'
 import CountdownTimer from './CountdownTimer'
 import { SettingsProvider } from '../../Context/SettingsContext'
 import { ViewProvider } from '../../Context/ViewContext'
+import { getSettings } from '../../apiCalls'
+jest.mock('../../apiCalls.js')
+
+describe('CountdownTimer', () => {
```
Comment (sienna-kopf) [inline src/Components/CountdownTimer/CountdownTimer.test.js:12]: Sweet test for the timer.

**Review by sienna-kopf — APPROVED** [review]
Super impressed with the tests you added for this @rwilliams659

File: package.json (hunk 1 above)
Comment (Abdeboskey) [reply package.json:24]: It seems like wizardry a little bit. Stoked that we're analyzing our coverage in this way!

File: src/Components/App/App.js [hunk 4]
```diff
@@ -9,30 +7,20 @@ import About from '../About/About'
...
 function App() {
-  if (!('Notification' in window)) {
-    console.log('This browser does not support notifications!')
-    return;
-  }
-
-  Notification.requestPermission(status => {
-    console.log('Notification permission status:', status)
-  })
+  enableNotifications()
```
Comment (Abdeboskey) [inline src/Components/App/App.js:13]: That's way cleaner with this coming in from a helper file.

File: src/Components/ContentDelivery/ContentDelivery.js [hunk 5]
```diff
@@ -35,7 +35,7 @@ const ContentDelivery = () => {
       </section>
       { done && 
         <div className={style.workMsgModal}>
-          <h2 className={style.modalContent}>Let's get back to work!</h2>
+          <h2 className={style.modalContent}>Time to focus again!</h2>
```
Comment (Abdeboskey) [inline src/Components/ContentDelivery/ContentDelivery.js:38]: Awesome, thank you for changing this!

File: src/Components/CountdownTimer/CountdownTimer.js [hunk 6]
```diff
@@ -4,6 +4,7 @@ import { ViewContext } from '../../Context/ViewContext'
 import Timer from 'react-compound-timer'
 import style from './CountdownTimer.module.scss'
 import playAlertSound from '../../helpers/audioHelper'
+import { displayNotification } from '../../helpers/notificationHelpers'
```
Comment (Abdeboskey) [inline src/Components/CountdownTimer/CountdownTimer.js:7]: Helper files! So nice we did it thrice!!

File: src/Components/App/App.test.js [hunk 7]
```diff
@@ -1,9 +1,409 @@
...
+
+  //   fireEvent.click(yogaBtn)
+
+  //   const videoHeading = await waitFor(() => getByRole('heading', { name: /enjoy your break!/i }))
+
+    // const timerHeading = getByRole('heading', { name: /click ▶ to begin focusing/i })
+
+    // expect(timerHeading).toBeInTheDocument()
+  // })
 })
```
Comment (Abdeboskey) [inline src/Components/App/App.test.js:409]: This is a gigantic test file, way to go Rachel.

File: src/Components/CountdownTimer/CountdownTimer.test.js [hunk 8]
```diff
@@ -24,4 +39,34 @@ describe('Timer', () => {
...
+        </ViewProvider>
+      </SettingsProvider>
+    )
+
+    const startButton = await waitFor(() => getByRole('button', { name: /start/i }))
+    
+    fireEvent.click(startButton)
+
+    setTimeout(() => {
+      expect(playAlertSound).toHaveBeenCalledTimes(1)
```
Comment (Abdeboskey) [inline src/Components/CountdownTimer/CountdownTimer.test.js:67]: Whoa dang, this worked?!

**Review by Abdeboskey — APPROVED** [review]
This is awesome, thank you so much for doing all this Rachel!

**Review by Chulstro — APPROVED** [review]
Love the test coverage and how thorough this has been

**Review by jkwest-93 — COMMENTED** [review]
Great work here, Rachel!! Our test coverage is looking great. I'm looking forward to hearing how we should handle our custom service worker code.

**Pull request merged**